src/receiver.rs

```diff
@@ -1,199 +1,159 @@
-        // deleting it and moving on to the next file.
-        let fslock = &mut (*syn);
-
-        while fslock.writes_to_read > 0 {
-            fslock.receiver_read_id = fslock.receiver_read_id.wrapping_add(1);
-
-            if fslock.receiver_idx.is_none() {
-                fslock.receiver_idx = Some(fslock.write_bound.expect("NO BOUND"));
-            }
-            if fslock.receiver_idx.unwrap() < fslock.in_memory_idx {
-                let event = fslock
-                    .mem_buffer
-                    .pop_front()
-                    .expect("there was not an event in the in-memory");
-                fslock.writes_to_read -= 1;
-                fslock.receiver_idx = fslock.receiver_idx.map(|x| x + 1);
-                return Some(event);
-            } else if (fslock.disk_writes_to_read == 0)
-                && (fslock.receiver_idx.unwrap() >= fslock.in_memory_idx)
-            {
-                let event = fslock
-                    .disk_buffer
-                    .pop_front()
-                    .expect("there was not an event in the disk buffer!");
-                fslock.writes_to_read -= 1;
-                fslock.receiver_idx = fslock.receiver_idx.map(|x| x + 1);
-                return Some(event);
-            } else {
-                match self.fp.read_exact(&mut sz_buf) {
-                    Ok(()) => {
-                        let payload_size_in_bytes = u8tou32abe(&sz_buf);
-                        let mut payload_buf = vec![0; (payload_size_in_bytes as usize)];
-                        match self.fp.read_exact(&mut payload_buf) {
-                            Ok(()) => match deserialize(&payload_buf) {
+    // This function is _only_ called when there's disk writes to be read. If a
+    // disk read happens and no `T` is returned this is an unrecoverable error.
+    fn read_disk_value(&mut self) -> Result<T, super::Error> {
+        loop {
+            match self.fp.read_u64::<BigEndian>() {
+                Ok(payload_size_in_bytes) => {
+                    let mut payload_buf = vec![0; payload_size_in_bytes as usize];
+                    match self.fp.read_exact(&mut payload_buf[..]) {
+                        Ok(()) => {
+                            let mut dec = DeflateDecoder::new(&payload_buf[..]);
+                            match deserialize_from(&mut dec, Infinite) {
                                 Ok(event) => {
-                                    fslock.receiver_idx = fslock.receiver_idx.map(|x| x + 1);
-                                    fslock.writes_to_read -= 1;
-                                    fslock.disk_writes_to_read -= 1;
-                                    return Some(event);
+                                    self.disk_writes_to_read -= 1;
+                                    return Ok(event);
                                 }
                                 Err(e) => panic!("Failed decoding. Skipping {:?}", e),
```

Comment:
Error says skipping, but isn't this a crash?

Reply:
Heh, the crashing-est kind of skip.

Will correct the language.